components/core/src/EncodedVariableInterpreter.cpp

```diff
@@ -240,24 +208,79 @@ void EncodedVariableInterpreter::encode_and_add_to_dictionary (const string& mes
     // To avoid reallocating the logtype as we append to it, reserve enough space to hold the entire message
     logtype_dict_entry.reserve_constant_length(message.length());
     while (logtype_dict_entry.parse_next_var(message, var_begin_pos, var_end_pos, var_str)) {
-        // Encode variable
-        encoded_variable_t encoded_var;
-        if (convert_string_to_representable_integer_var(var_str, encoded_var)) {
-            logtype_dict_entry.add_int_var();
-        } else if (convert_string_to_representable_float_var(var_str, encoded_var)) {
-            logtype_dict_entry.add_float_var();
-        } else {
-            // Variable string looks like a dictionary variable, so encode it as so
-            variable_dictionary_id_t id;
-            var_dict.add_entry(var_str, id);
-            encoded_var = encode_var_dict_id(id);
-            var_ids.push_back(id);
+        auto encoded_var = encode_var(var_str, logtype_dict_entry, var_dict, var_ids);
+        encoded_vars.push_back(encoded_var);
+    }
+}
 
-            logtype_dict_entry.add_dictionary_var();
-        }
+template<typename encoded_variable_t>
+void EncodedVariableInterpreter::encode_and_add_to_dictionary(
+        ir::LogEvent<encoded_variable_t> const& log_event,
+        LogTypeDictionaryEntry& logtype_dict_entry,
+        VariableDictionaryWriter& var_dict,
+        std::vector<ffi::eight_byte_encoded_variable_t>& encoded_vars,
+        std::vector<variable_dictionary_id_t>& var_ids,
+        size_t& raw_num_bytes
+) {
+    logtype_dict_entry.clear();
+    logtype_dict_entry.reserve_constant_length(log_event.get_logtype().length());
+
+    raw_num_bytes = 0;
+
+    auto constant_handler = [&](std::string const& value, size_t begin_pos, size_t length) {
+        raw_num_bytes += length;
+        logtype_dict_entry.add_constant(value, begin_pos, length);
+    };
+
+    auto encoded_int_handler = [&](encoded_variable_t encoded_var) {
+        raw_num_bytes += ffi::decode_integer_var(encoded_var).length();
+        logtype_dict_entry.add_int_var();
 
+        ffi::eight_byte_encoded_variable_t eight_byte_encoded_var{};
+        if constexpr (std::is_same_v<encoded_variable_t, ffi::eight_byte_encoded_variable_t>) {
+            eight_byte_encoded_var = encoded_var;
+        } else {  // std::is_same_v<encoded_variable_t, ffi::four_byte_encoded_variable_t>
+            eight_byte_encoded_var = ffi::encode_four_byte_integer_as_eight_byte(encoded_var);
+        }
+        encoded_vars.push_back(eight_byte_encoded_var);
+    };
+
+    auto encoded_float_handler = [&](ffi::four_byte_encoded_variable_t encoded_var) {
+        raw_num_bytes += ffi::decode_float_var(encoded_var).length();
+        logtype_dict_entry.add_float_var();
+
+        ffi::eight_byte_encoded_variable_t eight_byte_encoded_var{};
+        if constexpr (std::is_same_v<encoded_variable_t, ffi::eight_byte_encoded_variable_t>) {
```

Comment:
In the other file, we have a an else if and else assert instead. do we have a guidline on when we should verify the template parameters' type and when we don't?

Reply:
Switched to defining template implementations in the cpp combined with explicit specializations.